### PR TITLE
[ios][av] Fix container default type to work cross-platform

### LINKED ISSUE #11952: Expo-AV: Recorded sounds on Android cannot be played on iOS
**Summary**

I am using the `Audio.Recording` class of `Expo-AV` to record sounds on my Chat App.
On **Android** when using the default audio preset config for `LOW_QUALITY`, it returns an audio file with the extension `.3gp`
And on **iOS** it returns an audio file with the extension.caf

The problem is that the Sound Player cannot play `.caf` files on **Android** and the same way, `.3gp` files cannot been played on **iOS**.

I also try to use the same `extension` using `.mp3, .mp4, .wav, ...` for both OS but, the result is the same: Audio files recorded on Android cannot be played on iOS, but the opposite is possible.

Is there any way to record audio files so that there can been played on both OS Platforms automatically ?

I have also try order customs recordingOptions, but nothing works for both OS

**Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!**

managed

**What platform(s) does this occur on?**

Android, iOS

**SDK Version (managed workflow only)**

40

**Environment**

Expo CLI 4.0.17 environment info:
    System:
      OS: macOS 11.0.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 10.16.0 - /usr/local/bin/node
      Yarn: 1.22.4 - ~/.yarn/bin/yarn
      npm: 6.14.4 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.2, DriverKit 20.0, macOS 11.0, tvOS 14.2, watchOS 7.1
      Android SDK:
        API Levels: 23, 25, 26, 27, 28, 29
        Build Tools: 23.0.1, 26.0.2, 27.0.3, 28.0.2, 28.0.3, 29.0.2
        System Images: android-16 | ARM EABI v7a, android-25 | Google APIs Intel x86 Atom_64, android-26 | Google APIs Intel x86 Atom, android-28 | Google Play Intel x86 Atom, android-29 | Google APIs Intel x86 Atom
    IDEs:
      Xcode: 12.2/12B45b - /usr/bin/xcodebuild
    npmPackages:
      expo: ^40.0.0 => 40.0.1
      react: 16.13.1 => 16.13.1
      react-dom: 16.13.1 => 16.13.1
      react-native: https://github.com/expo/react-native/archive/sdk-40.0.1.tar.gz => 0.63.2
      react-navigation: ~4.0.10 => 4.0.10
    npmGlobalPackages:
      expo-cli: 4.0.17
    Expo Workflow: managed

**Reproducible demo or steps to reproduce from a blank project**

Here is how I record the audio.
````
await Audio.requestPermissionsAsync();
            await Audio.setAudioModeAsync({
                allowsRecordingIOS: true,
                playsInSilentModeIOS: true,
            });
            console.log('Starting recording..');
            const recording = new Audio.Recording();

            const RECORDING_OPTIONS_PRESET_LOW_QUALITY: RecordingOptions = {
                android: {
                    extension: '.wav',
                    outputFormat: RECORDING_OPTION_ANDROID_OUTPUT_FORMAT_MPEG_4,
                    audioEncoder: RECORDING_OPTION_ANDROID_AUDIO_ENCODER_AMR_NB,
                    sampleRate: 44100,
                    numberOfChannels: 2,
                    bitRate: 128000,
                },
                ios: {
                    extension: '.wav',
                    audioQuality: RECORDING_OPTION_IOS_AUDIO_QUALITY_MIN,
                    sampleRate: 44100,
                    numberOfChannels: 2,
                    bitRate: 128000,
                    linearPCMBitDepth: 16,
                    linearPCMIsBigEndian: false,
                    linearPCMIsFloat: false,
                },
            };

            await recording.prepareToRecordAsync(RECORDING_OPTIONS_PRESET_LOW_QUALITY);
            await recording.startAsync();
            console.log({recording})
            this.setState({recording});
            recording.setProgressUpdateInterval(1000);
````

After that, the audio is saved in an Amazon S3 and can be reached via an url like https://example.file.wav

The play the sound in the App, I use the following code:

````
const {sound} = await Audio.Sound.createAsync(
                   {uri: 'https://example.file.wav', headers: {}},
                   undefined,
                   onRecordingStatusUpdate,
                   false
               );
               setSound(sound);
               const playbackStatus: AVPlaybackStatus = await sound.playAsync();
               setIsPlaying(true);
````

### PR DESCRIPTION
**Pull request by actuallymentor**
With these defaults, ios audio can't be played on android. This codec change results in the files of ios and android being cross-playable.

Resolves #11952
Tagging @jonsamp for a second pair of eyes.

# Why

iOS `caf` files with their default codec cannot be played on Android. This goes against the cross-platform ethos of Expo.

# How

By changing the defaults to use a codec that both platforms support.

# Test Plan

New defaults have been tested to work by multiple users, see the [corresponding issue](https://github.com/expo/expo/issues/11952).

For context, copy-paste from issue:

```js
// Below code runs in a function

const { ios, android } = Audio.RECORDING_OPTIONS_PRESET_HIGH_QUALITY
await recorder.prepareToRecordAsync( {
	android: android,
	ios: {
		...ios,
		extension: '.mp4',
		outputFormat: Audio.RECORDING_OPTION_IOS_OUTPUT_FORMAT_MPEG4AAC
	}

} )
```

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-av/src/Audio/Recording.ts [hunk 1]
```diff
@@ -119,7 +119,8 @@ export const RECORDING_OPTIONS_PRESET_HIGH_QUALITY: RecordingOptions = {
     bitRate: 128000,
   },
   ios: {
-    extension: '.caf',
+    extension: '.mp4',
```
Comment (mnightingale) [inline packages/expo-av/src/Audio/Recording.ts:122]:
```suggestion
    extension: '.m4a',
```
Suggest this should be consistent with Android since they'll both be mpeg-4 containers?

**Review by mnightingale — COMMENTED** [review]
Change should be made on both RECORDING_OPTIONS_PRESET_HIGH_QUALITY and RECORDING_OPTIONS_PRESET_LOW_QUALITY

Perhaps we should consider switching the default to AAC_HE, when I researched it for my app HEv1 was supposedly as good quality at 64k vs 128k.

Android can encode HE-AACv1 since Android 4.1 https://developer.android.com/guide/topics/media/media-formats

Snippet of options I use at the moment
```
android: {
  // ...
  extension: '.m4a',
  outputFormat: Audio.RECORDING_OPTION_ANDROID_OUTPUT_FORMAT_MPEG_4,
  audioEncoder: Audio.RECORDING_OPTION_ANDROID_AUDIO_ENCODER_HE_AAC,
  bitRate: 64000,
  // ...
},
ios: {
  // ...
  extension: '.m4a',
  outputFormat: Audio.RECORDING_OPTION_IOS_OUTPUT_FORMAT_MPEG4AAC_HE,
  audioQuality: Audio.RECORDING_OPTION_IOS_AUDIO_QUALITY_MIN,
  bitRate: 64000,
  // ...
},

**Comment by actuallymentor** [comment]
Your suggestions seem sane, but I have not tested them. As I was trial and erroring I noticed that some codecs did not behave as the docs seemed the indicate they should.

For testing: the expo go client and the built packages also behaved differently.

File: packages/expo-av/src/Audio/Recording.ts (hunk 1 above)
Comment (bbarthec) [reply packages/expo-av/src/Audio/Recording.ts:122]: Let's switch to `.mp4` as suggested to obtain platforms unity 👍

**Review by bbarthec — CHANGES_REQUESTED** [review]
Thank you for this contribution. After I read the issue it looks fine to me 👍 
Please switch to `.mp4` as the default cross-platform container type and add respective CHANGELOG entry (in `BREAKING CHANGES` section).
Also rebuild the `build` folder in the `expo-av` to fix the CI (see https://github.com/expo/expo/blob/master/CONTRIBUTING.md).

**Comment by actuallymentor** [comment]
@bbarthec added the changelog as requested. Let me know if anything else needs to be done :)

File: packages/expo-av/src/Audio/Recording.ts (hunk 1 above)
Comment (mnightingale) [reply packages/expo-av/src/Audio/Recording.ts:122]: Are you suggesting switch both to `.mp4` or did you mean both to `.m4a`?
Before this PR its Android `.m4a` and iOS: `.caf`

Comment (actuallymentor) [reply packages/expo-av/src/Audio/Recording.ts:122]: No I would keep `m4a` on android since that is what I tested to work (see the issue). I'm happy to make it `mp4` but do not have the bandwidth to test it.

Comment (mnightingale) [reply packages/expo-av/src/Audio/Recording.ts:122]: Sorry yeah that was directed at @bbarthec

I think we want them both to be `mp4` or `m4a` but I was unclear if @bbarthec was suggesting making iOS `m4a` to match the existing Android value or change both iOS and Android to `mp4`

Comment (bbarthec) [reply packages/expo-av/src/Audio/Recording.ts:122]: Sorry guys for this misunderstanding of mine 🙇 
I would be awesome if we could ship the same extension/format on both platforms. As there's `.m4a` on Android we should probably lean towards the same on iOS. Is is possible? If so, I would go with it 😉

Comment (actuallymentor) [reply packages/expo-av/src/Audio/Recording.ts:122]: @bbarthec I have not tested it so can't confirm it works and do not have the time to test.

We can change to `m4a` globally but I can't guarantee it works. If you can test it on 2 hardware devices we can safely do `m4a`. Otherwise using `m4a` and `mp4` should be fine. We currently have 2 extensions anyway.

Comment (actuallymentor) [reply packages/expo-av/src/Audio/Recording.ts:122]: If @mnightingale is on board I can amend the PR to include it.

Comment (mnightingale) [reply packages/expo-av/src/Audio/Recording.ts:122]: I've used m4a for iOS/Android for quite a while, there'll be no issue mp4 === m4a they are a common extension for the same container (MPEG-4).
`m4a` is commonly used when the file only contains an audio track, likewise `mp4` when it has both audio and video tracks.
Since we are dealing with audio recordings it makes sense to unify on `m4a`.

Per my other comment I think we could consider changing the codecs to AAC_HE and possibly dropping the bitrate.

TL;DR do `.m4a` for both the high and low presets

**Comment by actuallymentor** [comment]
Ok as per the conversation with @bbarthec and @mnightingale I changed the container to `m4a` on iOS as well.

The other suggestion of @bbarthec to change codec's I think is beyond the scope of this PR.

As it stands I think this is ready for merge.

**Review by bbarthec — APPROVED** [review]

**Comment by mnightingale** [comment]
It's been merged but the change should also apple to the low quality preset, I guess it'll need another PR?

https://github.com/expo/expo/blob/82901c13484792b64273f78599a445993fa46097/packages/expo-av/src/Audio/Recording.ts#L140-L157

**Comment by bbarthec** [comment]
@mnightingale, yeah, another PR is a good way to go 👍
